src/nftOwnersLoop/nft_owners.ts

```diff
@@ -0,0 +1,232 @@
+    });
+  } else {
+    const transactionAddresses: {
+      from_address: string;
+      to_address: string;
+    }[] = await fetchTransactionAddressesFromBlock(
+      allContracts,
+      lastOwnersBlock,
+      blockReference
+    );
+    transactionAddresses.forEach((wallet) => {
+      addresses.add(wallet.from_address.toLowerCase());
+      addresses.add(wallet.to_address.toLowerCase());
+    });
+  }
+
+  logger.info({
+    owners: allOwners.length.toLocaleString(),
+    addresses: addresses.size.toLocaleString()
+  });
+
+  if (addresses.size > 0) {
+    logger.info(`[CALCULATING DELTA...]`);
+    const ownersDelta = await getOwnersDelta(
+      blockReference,
+      addresses,
+      allOwners,
+      reset
+    );
+
+    logger.info({
+      owners: allOwners.length.toLocaleString(),
+      ownersDelta: ownersDelta.length.toLocaleString()
+    });
+
+    await persistNftOwners(addresses, ownersDelta, reset);
+    await consolidateNftOwners(addresses, reset);
+  } else {
+    logger.info(`[NO CHANGES]`);
+  }
+};
+
+function ownersMatch(o1: NFTOwner, o2: OwnedNft) {
+  if (o1.token_id != o2.token_id) return false;
+  if (!areEqualAddresses(o1.wallet, o2.wallet)) return false;
+  if (!areEqualAddresses(o1.contract, o2.contract)) return false;
+  return true;
+}
+
+export async function getOwnersDelta(
+  blockReference: number,
+  addresses: Set<string>,
+  allOwners: OwnedNft[],
+  reset: boolean
+) {
+  if (reset) {
+    return allOwners.map((o) => {
+      return {
+        wallet: o.wallet.toLowerCase(),
+        contract: o.contract.toLowerCase(),
+        token_id: o.token_id,
+        balance: o.balance,
+        block_reference: blockReference
+      };
+    });
+  }
+
+  const ownersDelta: NFTOwner[] = [];
+
+  addresses.forEach((address) => {
+    const addressOwned = allOwners.filter((o) =>
+      areEqualAddresses(o.wallet, address)
+    );
+    addressOwned.forEach((o) => {
+      ownersDelta.push({
+        wallet: o.wallet.toLowerCase(),
+        contract: o.contract.toLowerCase(),
+        token_id: o.token_id,
+        balance: o.balance,
+        block_reference: blockReference
+      });
+    });
+  });
+
+  return ownersDelta;
+}
+
+export async function consolidateNftOwners(
+  addresses: Set<string>,
+  reset?: boolean
+) {
+  if (reset) {
+    const owners = await fetchAllNftOwners();
+    addresses.clear();
+    owners.forEach((owner) => {
+      addresses.add(owner.wallet);
+    });
+  }
+
+  logger.info(
+    `[CONSOLIDATING OWNERS FOR ${addresses.size.toLocaleString()} WALLETS] : [RESET ${reset}]`
+  );
+
+  if (!addresses.size) {
+    logger.info(`[NO WALLETS TO CONSOLIDATE]`);
+    return;
+  }
+
+  const upsertDeltaMap = new Map<string, ConsolidatedNFTOwner[]>();
+  const deleteDelta = new Set<string>();
+
+  await Promise.all(
+    Array.from(addresses).map(async (address) => {
+      const consolidation = (
+        await fetchWalletConsolidationKeysViewForWallet([address])
+      )?.[0];
```

Comment:
i think this is because await fetchWalletConsolidationKeysViewForWallet([address]) returns an array, if the array is empty then i need the `?.` otherwise empty[0] wouldn't work

Reply:
I don't think theres a problem with `[][0]`. Try it in your browser console.

```
>   [][0]
<- undefined
```